demos/mqtt/mqtt_demo_connection_sharing/mqtt_demo_connection_sharing.c

```diff
@@ -0,0 +1,1312 @@
+        SubscriptionManager_RemoveCallback( DEMO_PRECIPITATION_TOPIC,
+                                            DEMO_PRECIPITATION_TOPIC_LENGTH );
+
+        /* Set the subscription list memory to zero . */
+        ( void ) memset( ( void * ) pSubscriptionList, 0x00, sizeof( pSubscriptionList ) );
+
+        /* Populate the array with topic filters to unsubscribe from.
+         * We will use a single UNSUBSCRIBE packet to unsubscribe from all
+         * topic filters. */
+        pSubscriptionList[ 0 ].pTopicFilter = DEMO_TEMPERATURE_TOPIC_FILTER;
+        pSubscriptionList[ 0 ].topicFilterLength = DEMO_TEMPERATURE_TOPIC_FILTER_LENGTH;
+        pSubscriptionList[ 1 ].pTopicFilter = DEMO_HUMIDITY_TOPIC;
+        pSubscriptionList[ 1 ].topicFilterLength = DEMO_HUMIDITY_TOPIC_LENGTH;
+        pSubscriptionList[ 2 ].pTopicFilter = DEMO_PRECIPITATION_TOPIC;
+        pSubscriptionList[ 2 ].topicFilterLength = DEMO_PRECIPITATION_TOPIC_LENGTH;
+
+        /* Unsubscribe from all topic filters of temperature, humidity and
+         * precipitation data. */
+        returnStatus = unsubscribeFromTopicFilters( &mqttContext,
+                                                    pSubscriptionList,
+                                                    sizeof( pSubscriptionList )
+                                                    / sizeof( MQTTSubscribeInfo_t ) );
+    }
+
+    /* Send an MQTT Disconnect packet over the already connected TCP socket.
+     * There is no corresponding response for the disconnect packet. After sending
+     * disconnect, client must close the network connection. */
+    if( mqttSessionEstablished == true )
+    {
+        LogInfo( ( "Disconnecting the MQTT connection with %.*s.",
+                   BROKER_ENDPOINT_LENGTH,
+                   BROKER_ENDPOINT ) );
+
+        if( returnStatus == EXIT_FAILURE )
+        {
+            /* Returned status is not used to update the local status as there
+             * were failures in demo execution. */
+            ( void ) disconnectMqttSession( &mqttContext );
+        }
+        else
+        {
+            returnStatus = disconnectMqttSession( &mqttContext );
+        }
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Entry point of demo.
+ *
+ * The example shown below uses MQTT APIs to send and receive MQTT packets
+ * over the TLS connection established using OpenSSL.
+ *
+ * The example is single threaded and uses statically allocated memory;
+ * it uses QOS1 and therefore implements a retransmission mechanism
+ * for Publish messages. Retransmission of publish messages are attempted
+ * when a MQTT connection is established with a session that was already
+ * present. All the outgoing publish messages waiting to receive PUBACK
+ * are resent in this demo. In order to support retransmission all the outgoing
+ * publishes are stored until a PUBACK is received.
```

Comment:
Unless I've missed something retransmission won't work across reboots, so you may want to mention this in the comments.

Reply:
I have removed the text about re-transmission as I removed all the re-transmission related code from this demo to avoid code duplication across our demo files